### PR TITLE
convert more things to nanoseconds

### PR DESCRIPTION
**Pull request by Dudemanguy**
Depends on #12371. Merely using `ppoll` wrecks havoc on wayland's event loop for some reason so I omitted that. No idea why that happens (epoll weirdness?). Maybe that can be revisited later.

### CODE REVIEW

**Comment by Dudemanguy** [comment]
Very cool, freebsd fails building for some reason but I can't see anything in the CI logs. In theory, it's supposed to support ppoll like linux so not sure what the issue would be...

**Comment by NiLuJe** [comment]
```
../osdep/poll_wrapper.h:11:43: error: unknown type name 'int64_t'; did you mean '__int64_t'?
````

Missing `<stdint.h>` include? (Might be pulled in by another incloude on Linux and/or GCC?)

**Comment by Dudemanguy** [comment]
>Need to expand the run section to see the logs.

Oh I feel stupid. I was blocking the javascript that is apparently now required to see that...

File: video/out/wayland_common.c [hunk 1]
```diff
@@ -2604,12 +2604,12 @@ void vo_wayland_wait_frame(struct vo_wayland_state *wl)
...
 
-void vo_wayland_wait_events(struct vo *vo, int64_t until_time_us)
+void vo_wayland_wait_events(struct vo *vo, int64_t until_time_ns)
 {
     struct vo_wayland_state *wl = vo->wl;
 
-    int64_t wait_us = until_time_us - mp_time_us();
-    int timeout_ms = MPCLAMP((wait_us + 999) / 1000, 0, 10000);
+    int64_t wait_ns = until_time_ns - mp_time_ns();
+    int timeout_ms = MPCLAMP((wait_ns + 999000) / 1e6, 0, 10000);
```
Comment (kasper93) [inline video/out/wayland_common.c:2612]: Should be `999999`. But why do we even round up the sleep time? Is there good reason for it, or just because?

**Comment by kasper93** [comment]
LGTM

File: video/out/wayland_common.c (hunk 1 above)
Comment (Dudemanguy) [reply video/out/wayland_common.c:2612]: Probably something cargoculted from the distant past I'm guessing.

Comment (Dudemanguy) [reply video/out/wayland_common.c:2612]: Ah avoiding a 0 wait apparently according to this https://github.com/mpv-player/mpv/commit/37d6604d70c8c594de2817db26356c4c950ac0fd. Could have just clamped it to 1 though? Also, I just noticed that drm was never changed so it still uses 500.

**Comment by Dudemanguy** [comment]
Also changed `mp_sleep_us` to `mp_sleep_ns` since that was trivial.

**Comment by kasper93** [comment]
> Also changed mp_sleep_us to mp_sleep_ns since that was trivial.

At this point you might also cherry-pick this commit for maximum memes https://github.com/mpv-player/mpv/pull/12045/commits/6883d1e27f1da96b6ea8569ecba6b53e81ebf5b3

Allows to request sleep with 100ns resolution. Although Windows kernel doesn't go below 0.5ms in practice, but we can defer this limitation to kernel and request high resolution from API.

I know @avih didn't like the additional code, but we are already bikeshedding the timer/timestamp code so might as well.

**Comment by Dudemanguy** [comment]
>LARGE_INTEGER time = (LARGE_INTEGER){ .QuadPart = -(us * 10) };

Sure why not.

File: osdep/timer-win2.c [hunk 2]
```diff
@@ -52,17 +52,41 @@ void mp_end_hires_timers(int res_ms)
...
+    // retry without it.
+    if (!timer)
+        timer = CreateWaitableTimerEx(NULL, NULL, 0, TIMER_ALL_ACCESS);
+
+    if (!timer)
+        goto end;
+
+    // Time is expected in 100 nanosecond intervals.
+    // Negative values indicate relative time.
+    LARGE_INTEGER time = (LARGE_INTEGER){ .QuadPart = -(ns * 10000) };
```
Comment (Dudemanguy) [inline osdep/timer-win2.c:80]: I feel like my conversion here is wrong. What's this supposed to be exactly?

Comment (Dudemanguy) [reply osdep/timer-win2.c]: Divided by 100 instead I think.

Comment (kasper93) [reply osdep/timer-win2.c]: Yes, `/ 100`

File: audio/out/ao_wasapi_utils.c [hunk 3]
```diff
@@ -945,7 +945,7 @@ bool wasapi_thread_init(struct ao *ao)
 {
     struct wasapi_state *state = ao->priv;
     MP_DBG(ao, "Init wasapi thread\n");
-    int64_t retry_wait = 1;
+    int64_t retry_wait = 1000;
```
Comment (kasper93) [inline audio/out/ao_wasapi_utils.c:948]: Not really important for this PR. But maybe we could have macros for defining wait times in more readable way

like 

```
#define MP_TIME_S(s) ((s) * 1000000000)
#define MP_TIME_MS(ms) ((ms) * 1000000)
#define MP_TIME_US(ms) ((ms) * 1000)
#define MP_TIME_NS(ns) ((ns))
```
then we can abstract all zeros and multiplications. and just do `MP_TIME_MS(1)`, especially we do a lot `MP_TIME_S(1)` conversions.

Comment (Dudemanguy) [reply audio/out/ao_wasapi_utils.c:948]: Amusingly, `MP_SECOND_US` already exists apparently. What an awful name.

**Comment by Dudemanguy** [comment]
Considerably more risky than #12371, so I apologize in advance.

**Pull request merged**